infrastructure/src/main/java/org/corfudb/infrastructure/BatchWriter.java

```diff
@@ -0,0 +1,126 @@
+package org.corfudb.infrastructure;
+
+import com.github.benmanes.caffeine.cache.CacheWriter;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.log.LogAddress;
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.runtime.exceptions.OverwriteException;
+
+import javax.annotation.Nonnull;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Created by m on 11/30/16.
+ */
+@Slf4j
+public class BatchWriter <K, V> implements CacheWriter<K, V> {
+
+    private StreamLog streamLog;
+    private BlockingQueue<WriteOperation> writeQueue;
+    final ExecutorService writerService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+            .setDaemon(false)
+            .setNameFormat("LogUnit-Write-Processor-%d")
+            .build());
+
+    public BatchWriter(StreamLog streamLog) {
+        this.streamLog = streamLog;
+        writeQueue = new LinkedBlockingQueue<>();
+        writerService.submit(this::batchWriteProcessor);
+    }
+
+    @Override
+    public void write(@Nonnull K key, @Nonnull V value) {
+        try {
+            submitWrite((LogAddress) key, (LogData) value).get();
+        } catch (Exception e) {
+            if(e.getCause() instanceof OverwriteException) {
+                throw new OverwriteException();
+            } else {
+                throw new RuntimeException(e);
```

Comment:
Let's add a WARN (or maybe higher?) level log statement here, since this would be an unexpected error.